sql/src/test/java/io/crate/planner/consumer/ManyTableConsumerTest.java

```diff
@@ -197,4 +200,63 @@ public void test3TableSortOnWhere() throws Exception {
         assertThat(t1Andt3.toString(), is("join.doc.t1.doc.t3"));
         assertThat(root.right().getQualifiedName().toString(), is("doc.t2"));
     }
+
+    private static final Map<QualifiedName, AnalyzedRelation> sources = ImmutableMap.of(
+        new QualifiedName(T3.T1_INFO.ident().name()), T3.TR_1,
+        new QualifiedName(T3.T2_INFO.ident().name()), T3.TR_2,
+        new QualifiedName(T3.T3_INFO.ident().name()), T3.TR_3,
+        new QualifiedName(T3.T4_INFO.ident().name()), T3.TR_4);
+
+    private static final SqlExpressions expressions = new SqlExpressions(sources);
+
+    @Test
+    public void testReordering4TableSortOnWhereAndJoinConditionsThatDontMatchThePair() throws Exception {
+        List<QualifiedName> relations = ImmutableList.of(
+            T3.T1,
+            T3.T2,
+            T3.T3,
+            T3.T4);
+        List<JoinPair> joinPairs = ImmutableList.of(
+            new JoinPair(T3.T1, T3.T2, JoinType.INNER, expressions.asSymbol("t1.a=t2.b")),
+            new JoinPair(T3.T2, T3.T3, JoinType.INNER, expressions.asSymbol("t2.b=t1.a")),
+            new JoinPair(T3.T4, T3.T3, JoinType.INNER, expressions.asSymbol("t4.id=t3.c")));
+        assertThat(ManyTableConsumer.orderByJoinConditions(
+            relations,
+            Collections.emptySet(),
+            joinPairs,
+            ImmutableList.of(T3.T4)), contains(T3.T4, T3.T1, T3.T3, T3.T2));
+    }
+
+    @Test
+    public void test4TableSortOnWhereAndJoinConditionsThatDontMatchThePair() throws Exception {
+        MultiSourceSelect mss = analyze("select *" +
+                                        " from t1" +
+                                        " join t2 on t1.a=t2.b" +
+                                        " join t3 on t2.b=t3.c" +
+                                        " join users on t1.i=users.id" +
+                                        " join users_multi_pk on t3.z=users_multi_pk.id" +
+                                        " order by t3.c");
+        TwoTableJoin root = ManyTableConsumer.buildTwoTableJoinTree(mss);
+        assertThat(root.toString(), is("join.join.join.join.doc.t3.doc.t1.doc.users_multi_pk.doc.t2.doc.users"));
```

Comment:
probably out of scope for this PR, but maybe we could come up with a better way to represent two-table-joins? This is very difficult to read